mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+	if txmp.IsRejectedTx(key) {
+		// The peer has sent us a transaction that we have previously marked as invalid. Since `CheckTx` can
+		// be non-deterministic, we don't punish the peer but instead just ignore the tx
+		return nil, ErrTxAlreadyRejected
+	}
+
+	if txmp.WasRecentlyEvicted(key) {
+		// the transaction was recently evicted. If true, we attempt to re-add it to the mempool
+		// skipping check tx.
+		return txmp.TryReinsertEvictedTx(key, tx, txInfo.SenderID)
+	}
+
+	if txmp.Has(key) {
+		txmp.metrics.AlreadySeenTxs.Add(1)
+		// The peer has sent us a transaction that we have already seen
+		return nil, ErrTxInMempool
+	}
+
+	// reserve the key
+	if !txmp.store.reserve(key) {
+		txmp.logger.Debug("mempool already attempting to verify and add transaction", "txKey", fmt.Sprintf("%X", key))
+		txmp.PeerHasTx(txInfo.SenderID, key)
+		return nil, ErrTxInMempool
+	}
+	defer txmp.store.release(key)
+
+	// If a precheck hook is defined, call it before invoking the application.
+	if err := txmp.preCheck(tx); err != nil {
+		txmp.metrics.FailedTxs.Add(1)
+		return nil, mempool.ErrPreCheck{Reason: err}
+	}
+
+	// Early exit if the proxy connection has an error.
+	if err := txmp.proxyAppConn.Error(); err != nil {
+		return nil, err
+	}
+
+	// Invoke an ABCI CheckTx for this transaction.
+	rsp, err := txmp.proxyAppConn.CheckTxSync(abci.RequestCheckTx{Tx: tx})
+	if err != nil {
+		return rsp, err
+	}
+	if rsp.Code != abci.CodeTypeOK {
+		txmp.rejectedTxCache.Push(key)
+		txmp.metrics.FailedTxs.Add(1)
+		return rsp, fmt.Errorf("application rejected transaction with code %d and log %s", rsp.Code, rsp.Log)
+	}
+
+	// Create wrapped tx
+	wtx := newWrappedTx(
+		tx, key, txmp.Height(), rsp.GasWanted, rsp.Priority, rsp.Sender,
+	)
+
+	// Perform the post check
+	err = txmp.postCheck(wtx.tx, rsp)
+	if err != nil {
+		txmp.rejectedTxCache.Push(key)
+		txmp.metrics.FailedTxs.Add(1)
+		return rsp, fmt.Errorf("rejected bad transaction after post check: %w", err)
+	}
+
+	// Now we consider the transaction to be valid. Once a transaction is valid, it
+	// can only become invalid if recheckTx is enabled and RecheckTx returns a non zero code
+	if err := txmp.addNewTransaction(wtx, rsp); err != nil {
+		return nil, err
+	}
+	return rsp, nil
+}
+
+// RemoveTxByKey removes the transaction with the specified key from the
+// mempool. It adds it to the rejectedTxCache so it will not be added again
+func (txmp *TxPool) RemoveTxByKey(txKey types.TxKey) error {
+	txmp.removeTxByKey(txKey)
+	txmp.metrics.EvictedTxs.Add(1)
+	return nil
+}
+
+func (txmp *TxPool) removeTxByKey(txKey types.TxKey) {
+	txmp.rejectedTxCache.Push(txKey)
+	_ = txmp.evictedTxs.Pop(txKey)
+	_ = txmp.store.remove(txKey)
+	txmp.seenByPeersSet.RemoveKey(txKey)
+	txmp.metrics.EvictedTxs.Add(1)
+	txmp.broadcastMtx.Lock()
+	defer txmp.broadcastMtx.Unlock()
+	txmp.txsToBeBroadcast = make(map[types.TxKey]struct{})
```

Comment:
why does removing a single transaction completely reset this map? Can't their be other txns in this map pending broadcast?

Reply:
Yeah that's wrong. Good catch! It should just be removing that transaction from there (if it exists)